### DISCUSSION TITLE
specify hip-runtime-amd instead of hip for deps

### DISCUSSION THREAD

**Discussion by acxz** [rocm-arch/rocm-arch#753] [Ideas]
Basically I'd like to propose for the next patch release, all instances of hip be converted to hip-runtime-amd in the deps.

the hip* packages should be treated as the miopen- packages where there is miopen-hip and miopen-opencl but in hip's case it should be say hipsparse or hipsparse-cuda.

How does this sound @tpkessler @JP-Ellis ?

As it is right now users are allowed to build hip* packages with both hip-runtime-amd and hip-runtime-nvidia but many of these will fail without an explicit cuda dep and with some flags like BUILD_CUDA for hipsparse.

This may not affect a lot of people in the first place since I doubt many nvidia users are utilizing hip currently, however, it is still functionality that should exist at some point. We shouldn't worry about creating the hip*-cuda packages for now. (I'll get around to this eventually) Just making sure people don't accidentally try to build these packages with the nvidia backend is a good solution for now.

Note certain packages should be exclude from this, this includes the `miopen-hip` package.

**Comment by tpkessler** [comment]
I agree. I've got a couple of questions: 

* How does AMD handle this?
* What is best practice in arch packaging for marshalling libraries with different backends?
* How would we maintain the CUDA related packages? I don't own an nvidia card and don't want to install CUDA.

**Comment by acxz** [comment]
> How does AMD handle this?

From what I am aware, AMD does not handle this at all. They allow you to build the CUDA backend from source, but do not provide binaries with associated CUDA compiled code. I'll try to find further proof on this and see if in their build scripts how they handle it.

> What is best practice in arch packaging for marshalling libraries with different backends?

I have seen a couple methods around.
1. the method we have with `miopen-hip/opencl` you generate a separate package with the different backend
2. using optdepends if the package allows for it

I would prefer the optdeps method as that eases the maintenance burden, but can see it getting ugly for users if we dont have a hard req on either backend. So personally vote is on method 1)

> How would we maintain the CUDA related packages? I don't own an nvidia card and don't want to install CUDA.

What if I told you that I have never owned an AMD card? Ever since I started this project in early 2020 I have never tested the packages here with AMD hardware. I use CUDA heavily for my research and would be willing to maintain the CUDA related packages.

**Comment by JP-Ellis** [comment]
From my own observations, I have noticed optional dependencies being used where a package can easily switch between backends, or where a dependency provides an additional optional features.  In cases where the builds are quite different though, I am not so sure. I think there is a case for having different package names from looking at `tensorflow`, `tensorflow-cuda`, `tensorflow-opt` and `tensorflow-opt-cuda`.
